quarkus.ls/com.redhat.quarkus.ls/src/main/java/com/redhat/quarkus/services/QuarkusCodeActions.java

```diff
@@ -107,6 +111,77 @@ private void doCodeActionsForUnknown(Diagnostic diagnostic, PropertiesModel docu
 		}
 	}
 
+	/**
+	 * Returns the <code>Position</code> to insert the missing required code action property into
+	 * @param textDocument the text document
+	 * @return the <code>Position</code> to insert the missing required code action property into
+	 * @throws BadLocationException
+	 */
+	private Position getPositionForRequiredCodeAction(TextDocument textDocument) throws BadLocationException {
+		String textDocumentText = textDocument.getText();
+
+		Position position;
+		
+		if (!StringUtils.hasText(textDocumentText)) {
+			position = new Position(0, 0);
+		} else {
+			position = textDocument.positionAt(StringUtils.trimEnd(textDocumentText).length());
+		}
+		return position;
+	}
+
+	/**
+	 * Returns the missing required property name from <code>diagnosticMessage</code>
+	 * @param diagnosticMessage the diagnostic message containing the property name in single quotes
+	 * @return the missing required property name from <code>diagnosticMessage</code>
+	 */
+	private String getPropertyNameFromRequiredMessage(String diagnosticMessage) {
+		String property = diagnosticMessage.substring(diagnosticMessage.indexOf("'") + 1);
+		return property.substring(0, property.indexOf("'"));
+	}
+
+	/**
+	 * Create a code action that inserts all missing required properties and equals signs if there
+	 * are more than one missing required properties.
+	 * 
+	 * @param diagnostics        the diagnostics, one for each missing required property
+	 * @param document           the properties model
+	 * @param formattingSettings the formatting settings
+	 * @param codeActions        the code actions list to fill
+	 */
+	private void doCodeActionForAllRequired(List<Diagnostic> diagnostics, PropertiesModel document,
+			QuarkusFormattingSettings formattingSettings, List<CodeAction> codeActions) {
+		
+		TextDocument textDocument = document.getDocument();
+		List<String> properties = new ArrayList<>();
+		for (Diagnostic diagnostic : diagnostics) {
+			if (ValidationType.required.name().equals(diagnostic.getCode())) {
+				properties.add(getPropertyNameFromRequiredMessage(diagnostic.getMessage()) + (formattingSettings.isSurroundEqualsWithSpaces() ? " = " : "="));
+			}
+		}
+
+		if (properties.isEmpty()) {
+			return;
+		}
+
+		try {
+			Position position = getPositionForRequiredCodeAction(textDocument);
+			String stringToInsert = StringUtils.join(properties, "\n");

```

Comment:
\n is working for Linux but not for windows OS. You should use delemiter from the first line of the text document (see the following code that I have not tested).

After that you don't need properties list, you can directly create the stringToInsert (which should come a StringBuilder) by looping from requiredDiagnostics.

```java
String lineDelimiter = document.getDocument().lineDelimiter(0);
String assign = formattingSettings.isSurroundEqualsWithSpaces() ? " = " : "=";
StringBuilder stringToInsert = new StringBuilder();
if (StringUtils.hasText(textDocument.getText())) {
	stringToInsert.append(lineDelimiter);
}
for (Diagnostic diagnostic : requiredDiagnostics) {
	stringToInsert.append(getPropertyNameFromRequiredMessage(diagnostic.getMessage()));
	stringToInsert.append(assign);
}
			
CodeAction insertAction = CodeActionFactory.insert(
		"Add all missing required properties?", position,
				stringToInsert.toString(), textDocument, requiredDiagnostics);
```

Reply:
I decided to change the for loop to:

```
for (int i = 0; i < requiredDiagnostics.size(); i++) {
	Diagnostic diagnostic = requiredDiagnostics.get(i);
	stringToInsert.append(getPropertyNameFromRequiredMessage(diagnostic.getMessage()));
	stringToInsert.append(assign);

	if (i < requiredDiagnostics.size() - 1) {
		stringToInsert.append(lineDelimiter);
	}
}
```

because it would be nice to not have a line delimiter after the last required property. What do you think?

Reply:
You are right @xorye my suggestion was just to explain my idea but I didn't test it and miss line delimiter.